File: backend/lib/__tests__/storage.test.js
const { createStorage, createStorageFromEnv, LocalStorage } = require('../storage');
const fs = require('fs').promises;
const path = require('path');
const os = require('os');

describe('Storage', () => {
  let tmpDir;

  beforeEach(async () => {
    tmpDir = await fs.mkdtemp(path.join(os.tmpdir(), 'storage-test-'));
  });

  afterEach(async () => {
    try {
      await fs.rm(tmpDir, { recursive: true, force: true });
    } catch (err) {
      // Ignore
    }
  });

  describe('LocalStorage', () => {
    let storage;

    beforeEach(async () => {
      storage = new LocalStorage(tmpDir);
      await storage.init();
    });

    it('should create storage instance', () => {
      expect(storage).toBeDefined();
      expect(storage.basePath).toBe(tmpDir);
    });

    it('should require basePath', () => {
      expect(() => new LocalStorage()).toThrow('basePath is required');
    });

    it('should put and get file', async () => {
      const key = 'test/file.txt';
      const content = Buffer.from('Hello, World!');

      await storage.put(key, content, content.length, 'text/plain');

      const stream = await storage.get(key);
      const chunks = [];
      for await (const chunk of stream) {
        chunks.push(chunk);
      }
      const result = Buffer.concat(chunks);

      expect(result.toString()).toBe('Hello, World!');
    });

    it('should check if file exists', async () => {
      const key = 'test/exists.txt';
      const content = Buffer.from('test');

      let exists = await storage.exists(key);
      expect(exists).toBe(false);

      await storage.put(key, content, content.length, 'text/plain');

      exists = await storage.exists(key);
      expect(exists).toBe(true);
    });

    it('should delete file', async () => {
      const key = 'test/delete.txt';
      const content = Buffer.from('to be deleted');

      await storage.put(key, content, content.length, 'text/plain');
      expect(await storage.exists(key)).toBe(true);

      await storage.delete(key);
      expect(await storage.exists(key)).toBe(false);
    });

    it('should list files', async () => {
      await storage.put('test/list/file1.txt', Buffer.from('test'), 4, 'text/plain');
      await storage.put('test/list/file2.txt', Buffer.from('test'), 4, 'text/plain');
      await storage.put('test/list/sub/file3.txt', Buffer.from('test'), 4, 'text/plain');

      const keys = await storage.list('test/list');
      expect(keys).toHaveLength(3);
      expect(keys).toContain('test/list/file1.txt');
      expect(keys).toContain('test/list/file2.txt');
      expect(keys).toContain('test/list/sub/file3.txt');
    });

    it('should generate file URL', async () => {
      const key = 'test/url.txt';
      const content = Buffer.from('test');

      await storage.put(key, content, content.length, 'text/plain');

      const url = await storage.url(key, 3600);
      expect(url).toMatch(/^file:\/\//);
      expect(url).toContain('url.txt');
    });

    it('should throw error for URL of non-existent file', async () => {
      await expect(storage.url('nonexistent.txt', 3600)).rejects.toThrow('File not found');
    });

    it('should stream file with range', async () => {
      const key = 'test/stream.txt';
      const content = Buffer.from('0123456789');

      await storage.put(key, content, content.length, 'text/plain');

      // Stream bytes 5-9
      const stream = await storage.stream(key, 5, 5);
      const chunks = [];
      for await (const chunk of stream) {
        chunks.push(chunk);
      }
      const result = Buffer.concat(chunks);

      expect(result.toString()).toBe('56789');
    });

    it('should stream file from offset to end', async () => {
      const key = 'test/stream2.txt';
      const content = Buffer.from('0123456789');

      await storage.put(key, content, content.length, 'text/plain');

      // Stream from offset 5 to end
      const stream = await storage.stream(key, 5, 0);
      const chunks = [];
      for await (const chunk of stream) {
        chunks.push(chunk);
      }
      const result = Buffer.concat(chunks);

      expect(result.toString()).toBe('56789');
    });

    it('should handle nested directories', async () => {
      const key = 'a/b/c/d/e/file.txt';
      const content = Buffer.from('nested');

      await storage.put(key, content, content.length, 'text/plain');

      const exists = await storage.exists(key);
      expect(exists).toBe(true);
    });

    it('should handle empty content', async () => {
      const key = 'empty.txt';
      const content = Buffer.from('');

      await storage.put(key, content, 0, 'text/plain');

      const stream = await storage.get(key);
      const chunks = [];
      for await (const chunk of stream) {
        chunks.push(chunk);
      }
      const result = Buffer.concat(chunks);

      expect(result.length).toBe(0);
    });

    it('should not error when deleting non-existent file', async () => {
      await expect(storage.delete('nonexistent.txt')).resolves.not.toThrow();
    });

    it('should return empty array for non-existent prefix', async () => {
      const keys = await storage.list('nonexistent/prefix');
      expect(keys).toEqual([]);
    });
  });

  describe('createStorage', () => {
    it('should create local storage', async () => {
      const storage = await createStorage({
        backend: 'local',
        localPath: tmpDir,
      });

      expect(storage).toBeInstanceOf(LocalStorage);
    });

    it('should default to local storage', async () => {
      const storage = await createStorage({
        localPath: tmpDir,
      });

      expect(storage).toBeInstanceOf(LocalStorage);
    });

    it('should throw for unknown backend', async () => {
      await expect(
        createStorage({
          backend: 'unknown',
          localPath: tmpDir,
        })
      ).rejects.toThrow('Unknown storage backend');
    });
  });

  describe('createStorageFromEnv', () => {
    let originalEnv;

    beforeEach(() => {
      originalEnv = { ...process.env };
    });

    afterEach(() => {
      process.env = originalEnv;
    });

    it('should create storage from environment variables', async () => {
      process.env.STORAGE_BACKEND = 'local';
      process.env.STORAGE_LOCAL_PATH = tmpDir;

      const storage = await createStorageFromEnv();

      expect(storage).toBeInstanceOf(LocalStorage);
      expect(storage.basePath).toBe(tmpDir);
    });

    it('should default to local backend', async () => {
      delete process.env.STORAGE_BACKEND;
      process.env.STORAGE_LOCAL_PATH = tmpDir;

      const storage = await createStorageFromEnv();

      expect(storage).toBeInstanceOf(LocalStorage);
    });

    it('should default to /var/lib/nself-tv/media for local path', async () => {
      delete process.env.STORAGE_LOCAL_PATH;

      const storage = await createStorageFromEnv();

      expect(storage).toBeInstanceOf(LocalStorage);
      expect(storage.basePath).toBe('/var/lib/nself-tv/media');
    });
  });
});

File: backend/pkg/storage/storage_test.go
package storage

import (
	"bytes"
	"context"
	"io"
	"os"
	"path/filepath"
	"strings"
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestLocalStorage(t *testing.T) {
	// Create temp directory for tests
	tmpDir := t.TempDir()
	storage, err := NewLocalStorage(tmpDir)
	require.NoError(t, err)
	require.NotNil(t, storage)

	ctx := context.Background()

	t.Run("Put and Get", func(t *testing.T) {
		key := "test/file.txt"
		content := []byte("Hello, World!")

		err := storage.Put(ctx, key, bytes.NewReader(content), int64(len(content)), "text/plain")
		assert.NoError(t, err)

		reader, err := storage.Get(ctx, key)
		require.NoError(t, err)
		defer reader.Close()

		data, err := io.ReadAll(reader)
		assert.NoError(t, err)
		assert.Equal(t, content, data)
	})

	t.Run("Exists", func(t *testing.T) {
		key := "test/exists.txt"
		content := []byte("test")

		exists, err := storage.Exists(ctx, key)
		assert.NoError(t, err)
		assert.False(t, exists)

		err = storage.Put(ctx, key, bytes.NewReader(content), int64(len(content)), "text/plain")
		assert.NoError(t, err)

		exists, err = storage.Exists(ctx, key)
		assert.NoError(t, err)
		assert.True(t, exists)
	})

	t.Run("Delete", func(t *testing.T) {
		key := "test/delete.txt"
		content := []byte("to be deleted")

		err := storage.Put(ctx, key, bytes.NewReader(content), int64(len(content)), "text/plain")
		assert.NoError(t, err)

		err = storage.Delete(ctx, key)
		assert.NoError(t, err)

		exists, err := storage.Exists(ctx, key)
		assert.NoError(t, err)
		assert.False(t, exists)
	})

	t.Run("List", func(t *testing.T) {
		// Create multiple files
		keys := []string{
			"test/list/file1.txt",
			"test/list/file2.txt",
			"test/list/subdir/file3.txt",
		}
		for _, key := range keys {
			err := storage.Put(ctx, key, bytes.NewReader([]byte("test")), 4, "text/plain")
			assert.NoError(t, err)
		}

		// List all files with prefix
		results, err := storage.List(ctx, "test/list")
		assert.NoError(t, err)
		assert.Len(t, results, 3)
	})

	t.Run("URL", func(t *testing.T) {
		key := "test/url.txt"
		content := []byte("test")

		err := storage.Put(ctx, key, bytes.NewReader(content), int64(len(content)), "text/plain")
		assert.NoError(t, err)

		url, err := storage.URL(ctx, key, 1*time.Hour)
		assert.NoError(t, err)
		assert.True(t, strings.HasPrefix(url, "file://"))
		assert.Contains(t, url, "test/url.txt")
	})

	t.Run("Stream with offset", func(t *testing.T) {
		key := "test/stream.txt"
		content := []byte("0123456789")

		err := storage.Put(ctx, key, bytes.NewReader(content), int64(len(content)), "text/plain")
		assert.NoError(t, err)

		// Read bytes 5-9 (5 bytes starting at offset 5)
		reader, err := storage.Stream(ctx, key, 5, 5)
		require.NoError(t, err)
		defer reader.Close()

		data, err := io.ReadAll(reader)
		assert.NoError(t, err)
		assert.Equal(t, []byte("56789"), data)
	})

	t.Run("Stream without length", func(t *testing.T) {
		key := "test/stream2.txt"
		content := []byte("0123456789")

		err := storage.Put(ctx, key, bytes.NewReader(content), int64(len(content)), "text/plain")
		assert.NoError(t, err)

		// Read from offset 5 to end
		reader, err := storage.Stream(ctx, key, 5, 0)
		require.NoError(t, err)
		defer reader.Close()

		data, err := io.ReadAll(reader)
		assert.NoError(t, err)
		assert.Equal(t, []byte("56789"), data)
	})

	t.Run("Get non-existent file", func(t *testing.T) {
		_, err := storage.Get(ctx, "nonexistent/file.txt")
		assert.Error(t, err)
	})

	t.Run("Delete non-existent file", func(t *testing.T) {
		err := storage.Delete(ctx, "nonexistent/file.txt")
		assert.NoError(t, err) // Should not error
	})
}

func TestNewFactory(t *testing.T) {
	tmpDir := t.TempDir()

	t.Run("Create local storage", func(t *testing.T) {
		cfg := Config{
			Backend:   "local",
			LocalPath: tmpDir,
		}

		storage, err := New(cfg)
		assert.NoError(t, err)
		assert.NotNil(t, storage)

		// Verify it's a LocalStorage
		_, ok := storage.(*LocalStorage)
		assert.True(t, ok)
	})

	t.Run("Create S3 storage", func(t *testing.T) {
		cfg := Config{
			Backend:     "s3",
			S3Endpoint:  "http://localhost:9000",
			S3Region:    "us-east-1",
			S3AccessKey: "minioadmin",
			S3SecretKey: "minioadmin",
			S3Bucket:    "test-bucket",
			S3PathStyle: true,
		}

		storage, err := New(cfg)
		// May fail if MinIO not running, but should create the instance
		if err == nil {
			assert.NotNil(t, storage)
			_, ok := storage.(*S3Storage)
			assert.True(t, ok)
		}
	})

	t.Run("Create hybrid storage", func(t *testing.T) {
		cfg := Config{
			Backend:       "hybrid",
			LocalPath:     tmpDir,
			S3Endpoint:    "http://localhost:9000",
			S3Region:      "us-east-1",
			S3AccessKey:   "minioadmin",
			S3SecretKey:   "minioadmin",
			S3Bucket:      "test-bucket",
			S3PathStyle:   true,
			HybridWorkers: 2,
		}

		storage, err := New(cfg)
		// May fail if MinIO not running, but should create the instance
		if err == nil {
			assert.NotNil(t, storage)
			hybrid, ok := storage.(*HybridStorage)
			assert.True(t, ok)
			if hybrid != nil {
				hybrid.Close() // Clean up workers
			}
		}
	})

	t.Run("Unknown backend returns error", func(t *testing.T) {
		cfg := Config{
			Backend:   "unknown",
			LocalPath: tmpDir,
		}

		storage, err := New(cfg)
		assert.Error(t, err)
		assert.Nil(t, storage)
		assert.Contains(t, err.Error(), "unknown storage backend")
	})
}

func TestLocalStorageEdgeCases(t *testing.T) {
	tmpDir := t.TempDir()
	storage, err := NewLocalStorage(tmpDir)
	require.NoError(t, err)

	ctx := context.Background()

	t.Run("Create nested directories", func(t *testing.T) {
		key := "a/b/c/d/e/file.txt"
		content := []byte("nested")

		err := storage.Put(ctx, key, bytes.NewReader(content), int64(len(content)), "text/plain")
		assert.NoError(t, err)

		// Verify file exists
		fullPath := filepath.Join(tmpDir, key)
		_, err = os.Stat(fullPath)
		assert.NoError(t, err)
	})

	t.Run("List empty prefix", func(t *testing.T) {
		results, err := storage.List(ctx, "nonexistent/prefix")
		assert.NoError(t, err)
		assert.Empty(t, results)
	})

	t.Run("URL for non-existent file", func(t *testing.T) {
		_, err := storage.URL(ctx, "nonexistent.txt", 1*time.Hour)
		assert.Error(t, err)
	})

	t.Run("Stream non-existent file", func(t *testing.T) {
		_, err := storage.Stream(ctx, "nonexistent.txt", 0, 0)
		assert.Error(t, err)
	})

	t.Run("Empty content", func(t *testing.T) {
		key := "empty.txt"
		content := []byte("")

		err := storage.Put(ctx, key, bytes.NewReader(content), 0, "text/plain")
		assert.NoError(t, err)

		reader, err := storage.Get(ctx, key)
		require.NoError(t, err)
		defer reader.Close()

		data, err := io.ReadAll(reader)
		assert.NoError(t, err)
		assert.Empty(t, data)
	})
}

func TestNewLocalStorageCreatesDirectory(t *testing.T) {
	tmpDir := t.TempDir()
	newDir := filepath.Join(tmpDir, "new", "storage", "path")

	// Directory doesn't exist yet
	_, err := os.Stat(newDir)
	assert.True(t, os.IsNotExist(err))

	// NewLocalStorage should create it
	storage, err := NewLocalStorage(newDir)
	assert.NoError(t, err)
	assert.NotNil(t, storage)

	// Directory should now exist
	info, err := os.Stat(newDir)
	assert.NoError(t, err)
	assert.True(t, info.IsDir())
}
